### PR TITLE
Feat/tulip lending protocol

### PR DESCRIPTION
**Pull request by 00xSam**

### CODE REVIEW

File: ts-client/src/vault/constants.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import { Cluster, PublicKey } from '@solana/web3.js';
 import { BN } from '@project-serum/anchor';
 
-export const PROGRAM_ID = '24Uqj9JCLxUeoC3hGfh5W3s9FM9uCHDS2SG3LYwBpyTi';
+export const PROGRAM_ID = '6YRZW57XsrT2DxSNLXHHQd4QmiqBode4d6btASkRqcFo';
```
Comment (andrewsource147) [inline ts-client/src/vault/constants.ts:4]: Why we change program ID?

File: ts-client/src/vault/index.ts [hunk 2]
```diff
@@ -360,10 +360,14 @@ export default class VaultImpl implements VaultImplementation {
     const vaultStrategiesStatePromise = this.vaultState.strategies
       .filter((address) => address.toString() !== VAULT_STRATEGY_ADDRESS)
       .map(async (strat) => {
-        const strategyState = (await this.program.account.strategy.fetchNullable(strat)) as unknown as StrategyState;
+        const strategyState = (await this.program.account.strategy.fetch(strat)) as unknown as StrategyState;
         return { publicKey: strat, strategyState };
       });
     const vaultStrategiesState = await Promise.allSettled(vaultStrategiesStatePromise);
+    console.log(
```
Comment (andrewsource147) [inline ts-client/src/vault/index.ts:367]: Is it necessary for console log?

File: ts-client/src/vault/index.ts [hunk 3]
```diff
@@ -479,6 +483,9 @@ export default class VaultImpl implements VaultImplementation {
       withdrawOpt,
     );
 
+    console.log('lol', withdrawFromStrategyTx);
```
Comment (andrewsource147) [inline ts-client/src/vault/index.ts:486]: Ditto^

File: ts-client/src/vault/index.ts [hunk 4]
```diff
@@ -479,6 +483,9 @@ export default class VaultImpl implements VaultImplementation {
...
+    console.log('test', typeof withdrawFromStrategyTx);
+    console.log('test', withdrawFromStrategyTx instanceof Transaction);
```
Comment (andrewsource147) [inline ts-client/src/vault/index.ts:488]: Ditto^

File: ts-client/src/vault/strategy/tulip.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        affiliateId: PublicKey;
+        affiliateProgram: AffiliateVaultProgram;
+        partner: PublicKey;
+        user: PublicKey;
+      };
+    },
+  ): Promise<Transaction | { error: string }> {
+    if (!walletPubKey) throw new Error('No user wallet public key');
+
+    // https://github.com/Francium-DeFi/francium-sdk/blob/master/src/constants/lend/pools.ts#L59
```
Comment (andrewsource147) [inline ts-client/src/vault/strategy/tulip.ts:35]: Why francium comment in Tulip implementation?

File: ts-client/src/vault/strategy/tulip.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    const lendingPools = tulip.LENDING_RESERVES;
+    const lendingPool = Object.values(lendingPools).find(
+      (lendingPool: any) => lendingPool.account === strategy.state.reserve.toBase58(),
+    );
+    if (!lendingPool) throw new Error('Cannot find francium lending pool');
```
Comment (andrewsource147) [inline ts-client/src/vault/strategy/tulip.ts:40]: Ditto^

File: ts-client/src/vault/strategy/tulip.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        .accounts({
+          ...txAccounts,
+          partner: opt.affiliate.partner,
+          user: opt.affiliate.user,
+          vaultProgram: program.programId,
+          vaultLpMint: lpMint,
+          owner: walletPubKey,
+        })
+        .remainingAccounts(remainingAccounts)
+        .preInstructions(preInstructions)
```
Comment (andrewsource147) [inline ts-client/src/vault/strategy/tulip.ts:96]: We need preinstruction for refresh reserve
Ex: https://github.com/mercurial-finance/vault-sdk/blob/main/ts-client/src/vault/strategy/portWithoutLM.ts#L107

**Comment by andrewsource147** [comment]
Look good, but have you tested it?

**Comment by 00xSam** [comment]
> Look good, but have you tested it?

Yes, u can test it here https://mercurial-v2-git-tulip-integration-wowcats.vercel.app/?env=staging

**Review by andrewsource147 — APPROVED** [review]

**Pull request merged**